crates/polars-plan/src/plans/optimizer/cache_states.rs

```diff
@@ -290,13 +291,32 @@ pub(super) fn set_cache_states(
                         }
                     }
 
-                    let lp = lp_arena.take(node);
-                    let lp = proj_pd.optimize(lp, lp_arena, expr_arena)?;
-                    let lp = pred_pd.optimize(lp, lp_arena, expr_arena)?;
-                    lp_arena.replace(node, lp);
+                    let lp = lp_arena_scratch.take(node);
+                    let lp =
+                        proj_pd.optimize(lp, &mut lp_arena_scratch, &mut expr_arena_scratch)?;
+                    let lp =
+                        pred_pd.optimize(lp, &mut lp_arena_scratch, &mut expr_arena_scratch)?;
+                    lp_arena_scratch.replace(node, lp);
                 }
-                return Ok(());
+
+                let root_ir = IRNode::new(root);
+                let cmp_existing = root_ir
+                    .hashable_and_cmp(lp_arena, expr_arena)
+                    .ignore_caches();
+                let cmp_scratch = root_ir
+                    .hashable_and_cmp(&lp_arena_scratch, &expr_arena_scratch)
+                    .ignore_caches();
+                if cmp_existing != cmp_scratch {
+                    // If there was a change, we remove the cache nodes entirely
+                    if verbose {
+                        eprintln!("cache nodes will be removed because predicates don't match")
+                    }
+                    std::mem::swap(lp_arena, &mut lp_arena_scratch);
+                    std::mem::swap(expr_arena, &mut expr_arena_scratch);
+                }
+                continue;
```

Comment:
This was a return previously but I felt like a `continue` would be more appropriate. Please correct me if I'm wrong 👀